### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by yannickjmt**

### CODE REVIEW

File: wwwroot/assets/scripts/api.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import * as g from './global';
```
Comment (juanchaur) [inline wwwroot/assets/scripts/api.js]: improve naming.

File: wwwroot/assets/scripts/countries.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
+export default function () {
```
Comment (juanchaur) [inline wwwroot/assets/scripts/countries.js]: How about if you have as entity: Country then you just create a list of countries.

File: wwwroot/assets/scripts/global.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import indicatorsListMod from './indicatorsList';
```
Comment (juanchaur) [inline wwwroot/assets/scripts/global.js:1]: I see you are just ``` import * as g from 'global';```
I think it's better if you just export everything as a unique object.

File: wwwroot/assets/scripts/display.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+export const updateMapColors = () => {
+  // assign CSS class each to svg element if valid value
+  for (let svgCountry of g.svgCountries) {
+
+    let svgCountryCode = svgCountry.getAttribute('data-id');
+
+    let countryValue = g.countries.getValue(svgCountryCode, g.activeIndicator, g.activeYear);
+
+    let classCountry = (countryValue === '')
```
Comment (juanchaur) [inline wwwroot/assets/scripts/display.js]: as code standard, It's a bit weird to see it in 3 lines; And the parenthesis aren't needed.
```javascript
let classCountry = country === ''? 'noData' : getClassName(countryValue, g.activeIndicator, g.activeYear);
```

File: wwwroot/assets/scripts/display.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    $(textID).className = '';
+  }
+  $('#legend0').className = 'legendNoData';
+  $('#legend0-text').className = 'legend-text';
+  $('#legend0-text').classList.add('legend-text-noData');
+  $('#legend0-text').innerHTML = 'No data';
+};
+
+export const showCountryInfo = (event) => {
+  let countryCode = event.target.getAttribute('data-id');
```
Comment (juanchaur) [inline wwwroot/assets/scripts/display.js]: wouldn't it be better to have classes with those styles in it and just add the class needed?

File: wwwroot/assets/scripts/display.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+      if (g.activeIndicator != '') {
+        let value = g.countries.getValue(countryCode, g.activeIndicator, g.activeYear);
+        let indicatorName = g.legend.getIndicatorName(g.activeIndicator);
+
+        if (value === '') value = 'no data';
+        html += `<div>${indicatorName}: ${formatNumber(value)}</div>`;
+      }
+
+      tooltip.innerHTML = html;
+      tooltip.style.visibility = 'visible';
```
Comment (juanchaur) [inline wwwroot/assets/scripts/display.js]: same as above

File: wwwroot/assets/scripts/utils.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    return svg;
+  }
+  catch(err) {
+    log('error while retrieving svg: ' + err.message);
+  }
+};
+
+const ajax = (url, type) => {
+  return new Promise(function(resolve, reject) {
+    let xhr = new XMLHttpRequest();
```
Comment (juanchaur) [inline wwwroot/assets/scripts/utils.js]: why not using fetch?

File: wwwroot/assets/scripts/utils.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+};
+
+export const displaySpinner = (message) => {
+  $('#blank').style.display = 'block';
+  $('#spinner-text').innerHTML = message;
+  $('#spinner-wrapper').style.display = 'block';
+};
+
+export const hideSpinner = () => {
+  $('#spinner-text').innerHTML = '';
```
Comment (juanchaur) [inline wwwroot/assets/scripts/utils.js]: I think it's way easier to have state classes. Perhaps:
``` css
hidden {
    display: none;
}
```

File: wwwroot/assets/scripts/utils.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+  $('#spinner-wrapper').style.display = 'none';
+  $('#blank').style.display = 'none';
+};
+
+export const displayError = (message) => {
+  displaySpinner(message);
+  setTimeout( () => {
+    hideSpinner();
+  },2000);
```
Comment (juanchaur) [inline wwwroot/assets/scripts/utils.js]: magic number.

File: wwwroot/assets/scripts/utils.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+};
+
+export const formatNumber = (n) => {
+  if (typeof n == 'number') {
```
Comment (juanchaur) [inline wwwroot/assets/scripts/utils.js]: When you have more than 3 if/else is a smell that you need to split or rethink how to do it better.

File: wwwroot/assets/scripts/utils.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+      n = (Math.round(n / 100000000) / 10).toLocaleString('en-US') + 'b';
+    } else {
+      n = (Math.round(n / 1000000000)).toLocaleString('en-US') + 'b';
+    }
+  }
+  return n;
+};
+
+export const log = (message) => {
+  if (g.debug) {
```
Comment (juanchaur) [inline wwwroot/assets/scripts/utils.js]: perhaps you can have a window variable window.__debug set and that in the kick-off of the app if it's set to true would enable all console.logs

File: wwwroot/assets/scripts/utils.js [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+      else if (type == 'XML') resolve(this.responseXML);
+    };
+    xhr.onerror = reject;
+    xhr.open('GET', url);
+    xhr.send();
+  });
+};
+
+export const displaySpinner = (message) => {
+  $('#blank').style.display = 'block';
```
Comment (juanchaur) [inline wwwroot/assets/scripts/utils.js]: I think it's way easier to have state classes

File: wwwroot/assets/scripts/utils.js [hunk 13]
```diff
@@ -0,0 +1,73 @@
+const $ = document.querySelector.bind(document);
+import * as g from './global';
+
+export const getSVG = async (url, type) => {
+  try {
+    let svg = await ajax(url, type);
```
Comment (juanchaur) [inline wwwroot/assets/scripts/utils.js]: why not fetch?

File: wwwroot/assets/scripts/ui.js [hunk 14]
```diff
@@ -0,0 +1,178 @@
+const $ = document.querySelector.bind(document);
+var noUiSlider = require('nouislider');
```
Comment (juanchaur) [inline wwwroot/assets/scripts/ui.js]: why don't you use import? Try to be consistent

File: wwwroot/assets/scripts/ui.js [hunk 15]
```diff
@@ -0,0 +1,178 @@
+const $ = document.querySelector.bind(document);
```
Comment (juanchaur) [inline wwwroot/assets/scripts/ui.js]: all imports should be before any const.

File: wwwroot/assets/scripts/ui.js [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+  } 
+};
+
+const createUpdateSlider = (sliderElement, yearArr) => {
+  const filter = (value) => {
+    // filters value to display for scale
+    // can return 0 = no value, 1 = large value, 2 = small value
+    // type = 1 for min and max value, 2 for others
+    // this will display maximum 11 values on the slider
+    if (yearArr.length < 12) {
```
Comment (juanchaur) [inline wwwroot/assets/scripts/ui.js:135]: magic numbers

File: wwwroot/assets/scripts/ui.js [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+      return 1;
+    } else {
+      let tmp = Math.floor((yearArr.length + 8) / 10);
+      return value % tmp == 0 ? 1 : 0;
+    }
+  };
+
+  // cannot use noUiSlider.updateOptions() with current settings, so we need to destroy it
+  if (sliderElement.noUiSlider !== undefined) sliderElement.noUiSlider.destroy();
```
Comment (juanchaur) [inline wwwroot/assets/scripts/ui.js]:
```javascript
sliderElement.noUiSlider && sliderElement. noUiSlider.destoy();

File: wwwroot/assets/scripts/ui.js [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+const createUpdateSlider = (sliderElement, yearArr) => {
+  const filter = (value) => {
+    // filters value to display for scale
+    // can return 0 = no value, 1 = large value, 2 = small value
+    // type = 1 for min and max value, 2 for others
+    // this will display maximum 11 values on the slider
+    if (yearArr.length < 12) {
+      return 1;
+    } else {
+      let tmp = Math.floor((yearArr.length + 8) / 10);
```
Comment (juanchaur) [inline wwwroot/assets/scripts/ui.js]: same as above

File: wwwroot/assets/scripts/ui.js [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+
+  if ((g.activeYear == '') || (!yearArray.includes(g.activeYear))) {
+    g.activeYear = yearArray[0];
+  }
+
+  //can't create slider with only one value
+  //taken care of by forcing 2 years minimum range in the form
+  if (yearArray.length > 1) {
+    let slider = $('#slider');
+    createUpdateSlider(slider, yearArray);
```
Comment (juanchaur) [inline wwwroot/assets/scripts/ui.js]:
```javascript
createUpdateSlider($('#slider') , years);

File: wwwroot/assets/scripts/ui.js [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+  g.activeIndicator = s.options[s.selectedIndex].value;
+
+  // when indicator changes we may have to rebuild year slider
+  // because of indicator / year independance
+  buildYearsSelector();
+  fillMapAndLegend();
+};
+
+export const buildYearsSelector = () => {
+  let yearArray = g.legend.getYears(g.activeIndicator);
```
Comment (juanchaur) [inline wwwroot/assets/scripts/ui.js]: naming: years. 
It's plural so it's implicit that it's an list of year

File: wwwroot/assets/scripts/ui.js [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+  yearArray.sort((a, b) => a - b);
+
+  if ((g.activeYear == '') || (!yearArray.includes(g.activeYear))) {
```
Comment (juanchaur) [inline wwwroot/assets/scripts/ui.js]: parenthesis aren't needed.
```javascript
if( g.activeYear === ''  || !years.includes(g.activeYear) );

Comment (juanchaur) [inline wwwroot/assets/scripts/ui.js]: ```===```

File: wwwroot/assets/scripts/ui.js [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+    // select last indicator in list as active
+    // normally after new fetch data request, new indicator should be at end of array
+    g.activeIndicator = indicatorArray[indicatorArray.length - 1][0];
+
+    $('#indicators').addEventListener('change', listenChangeIndic);
+  }
+};
+
+const listenChangeIndic = () => {
+  let s = $('#indicators-select');
```
Comment (juanchaur) [inline wwwroot/assets/scripts/ui.js]: descriptive name

File: wwwroot/assets/scripts/paletteChange.js [hunk 23]
```diff
@@ -13,16 +13,17 @@ export default function (p) {
     ['#f7fcf0','#ccebe0','#9ed9cf','#77c6c1','#69adb7','#5b97ac','#4c80a1','#3b6997','#28548c','#084081'], // bluegreen blue
     ['#e0ffff','#d4e5ff','#c7ccff','#b8b2fd','#a999fa','#9781f4','#836beb','#6e55dd','#5942c9','#483d8b'], // LightCyan, Blue, DarkSlateBlue
     ['#ffffe0','#d6f3b7','#b0e393','#8fd375','#6fc259','#53af42','#3a9d2d','#238a1b','#0f760b','#006400'], // lightyellow, limegreen, DarkGreen
-    // ['#f7f4f9','#f4d4e8','#f0b3d6','#e992c6','#e270b5','#cd5798','#b44277','#9a2e58','#80193a','#67001f'], // pink purple redundant
     ['#fff7f3','#ffd5dd','#ffb0c8','#fc8bb2','#f1659f','#d05095','#ae3d8a','#8c2980','#6b1575','#49006a'], // pink-yellow purple
-
   ];
+
   p = (typeof p === 'undefined')
```
Comment (juanchaur) [inline wwwroot/assets/scripts/paletteChange.js]: one liner

File: wwwroot/assets/scripts/main.js [hunk 24]
```diff
@@ -2,80 +2,37 @@ import '../../../node_modules/nouislider/distribute/nouislider.css';
...
-import paletteChange from './paletteChange';
-
-var noUiSlider = require('nouislider');
+import * as g from './global';
+import paletteChangeMod from './paletteChange';
+import {getSVG, log} from './utils';
+import {showCountryInfo} from './display';
+import {buildForm} from './ui';
 
 'use strict';
```
Comment (juanchaur) [inline wwwroot/assets/scripts/main.js]: don't think is needed

File: wwwroot/assets/scripts/main.js [hunk 25]
```diff
@@ -85,543 +42,17 @@ document.addEventListener('DOMContentLoaded', function() {
   window.location.hash = 'modal';
 
   // country tooltip must follow mouse
-  document.addEventListener('mousemove', function(event) {
+  document.addEventListener('mousemove', (event) => {
     $('#tooltip').style.left = (event.pageX - 105) + 'px';
```
Comment (juanchaur) [inline wwwroot/assets/scripts/main.js]: magic numbers

File: wwwroot/assets/scripts/legend.js [hunk 26]
```diff
@@ -0,0 +1,111 @@
+export default function () {
+  let legend = {
```
Comment (juanchaur) [inline wwwroot/assets/scripts/legend.js]: same as with countries

File: wwwroot/assets/scripts/legend.js [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+
+  const reduceArray= (arr, rangeNum) => {
+    // create the legend range, will be used to separate the country values into equally large sets
+    // we get n+1 boundaries to build a legend with n colors (n = legendRangeNum)
+    const legendArr = [];
+    
+    // filter out null values
+    // we didn't filter at the country object level so that rest of the map could be drawn
+    // in case of whole set is null
+    arr = arr.filter(n => (n===null) ? false : true);
```
Comment (juanchaur) [inline wwwroot/assets/scripts/legend.js]: Descriptive naming
```javascript
arr = arr.filter(n => !n)

File: wwwroot/assets/scripts/legend.js [hunk 28]
```diff
@@ -0,0 +1,111 @@
...
+  
+    if (arr.length > 0) {
```
Comment (juanchaur) [inline wwwroot/assets/scripts/legend.js]: > 0 it's implicit
```javascript
if (arr.length)

File: wwwroot/assets/scripts/legend.js [hunk 29]
```diff
@@ -0,0 +1,111 @@
...
+      arr.sort((a, b) => a - b);
+      const max = Math.max(...arr);
+      for (let i = 0; i < rangeNum; i++) {
+        legendArr.push(arr[Math.round(i * arr.length / rangeNum)]);
```
Comment (juanchaur) [inline wwwroot/assets/scripts/legend.js]: I believe here you need to specify the operation order. In this case it's needed the parenthesis

File: wwwroot/assets/scripts/legend.js [hunk 30]
```diff
@@ -0,0 +1,111 @@
...
+
+    getIndicatorName(indicatorId) {
+      return this.indicators[indicatorId].indicator_name;
+    },
+
+    reduceLegend(rangeNum) {
+      for (let indicatorVal in this.indicators) {
+        for (let yearVal in this.indicators[indicatorVal]) {
+          let year = this.indicators[indicatorVal][yearVal];
+          if ((typeof year.reduced === 'boolean') && (year.reduced === false)) {
```
Comment (juanchaur) [inline wwwroot/assets/scripts/legend.js]: If you're checking the type perhaps it's better to rethink if that field should be split in 2 different ones.

File: wwwroot/assets/scripts/legend.js [hunk 31]
```diff
@@ -0,0 +1,111 @@
...
+      }
+      return null;
+    },
+
+    getYears(indicatorId) {
+      let yearArray = [];
+      // only get years relevant to active indicator
+      for (let yearVal in this.indicators[indicatorId]) {
+        let year = this.indicators[indicatorId][yearVal];
+        if (typeof year.reduced === 'boolean') {
```
Comment (juanchaur) [inline wwwroot/assets/scripts/legend.js]: If you're checking the type perhaps it's better to rethink if that field should be split in 2 different ones.

File: wwwroot/assets/scripts/legend.js [hunk 32]
```diff
@@ -0,0 +1,111 @@
...
+          //if (!yearArray.includes(yearVal)) yearArray.push(yearVal);
```
Comment (juanchaur) [inline wwwroot/assets/scripts/legend.js]: remove commented code

File: wwwroot/assets/scripts/legend.js [hunk 33]
```diff
@@ -0,0 +1,111 @@
...
+    //*              'values': [ , ]
+    //*             },
+    //*      Year2: { }
+    //*   }
+    //* }
+
+    indicators: {},
+
+    setValue(indicatorId, indicatorName, year, value) {
+      if (this.indicators[indicatorId]) {
```
Comment (juanchaur) [inline wwwroot/assets/scripts/legend.js]: Too many if else statements.

Comment (juanchaur) [reply wwwroot/assets/scripts/legend.js]: Split this method in documented private documented functions

File: wwwroot/assets/scripts/indicatorsList.js [hunk 34]
```diff
@@ -1,5 +1,7 @@
-export default function () {
+export default () => {
```
Comment (juanchaur) [inline wwwroot/assets/scripts/indicatorsList.js]: is it needed to be a function?
Can you just export the array?

**Review by juanchaur — CHANGES_REQUESTED** [review]

File: wwwroot/assets/scripts/api.js (hunk 1 above)
Comment (yannickjmt) [reply wwwroot/assets/scripts/api.js]: Is it really bad practice?
It's used everywhere, so I figured a shortcut like the $ for document.getElementBySomething would be ok.

You think all the g.xxx should become global.xxx for example?

File: wwwroot/assets/scripts/global.js (hunk 3 above)
Comment (yannickjmt) [reply wwwroot/assets/scripts/global.js:1]: You mean mix indicatorsList.js and global.js in one common object to export?

File: wwwroot/assets/scripts/ui.js (hunk 17 above)
Comment (yannickjmt) [reply wwwroot/assets/scripts/ui.js]: nice!

File: wwwroot/assets/scripts/ui.js (hunk 22 above)
Comment (yannickjmt) [reply wwwroot/assets/scripts/ui.js]: would indicatorChangeHandler be better?

File: wwwroot/assets/scripts/paletteChange.js (hunk 23 above)
Comment (yannickjmt) [reply wwwroot/assets/scripts/paletteChange.js]: Isn't it a bit hard to read now?
```javascript
p = (typeof p === 'undefined') ? 0 : pallettes[p+1] ? p+1 : 0;
```

File: wwwroot/assets/scripts/legend.js (hunk 29 above)
Comment (yannickjmt) [reply wwwroot/assets/scripts/legend.js]: I think it isn't for multiplications and divisions

File: wwwroot/assets/scripts/countries.js (hunk 2 above)
Comment (yannickjmt) [reply wwwroot/assets/scripts/countries.js]: I am not sure I follow you. The countries object is already a list of country entities.

**Comment by yannickjmt** [comment]
Wow that was very thorough! I am so grateful you took some times to check my code. Thanks a lot :)

I've almost done every changes you requested. What's the normal process normally, I push the commits and merge the PR?

File: wwwroot/assets/scripts/global.js (hunk 3 above)
Comment (juanchaur) [reply wwwroot/assets/scripts/global.js:1]: No I would export global as single object.
I know it's correct to export everything but If that's happening why not export an object with all those methods directly?

**Comment by juanchaur** [comment]
Yep, that's normally the process.
In my company we have checks that if the PR is not approved then you can't merge (in order to know that the code is correct and has been at least seen once by another team member)

**Review by juanchaur — APPROVED** [review]

**Pull request merged**